internal/signing/channel_account_db_signature_client.go

```diff
@@ -0,0 +1,108 @@
+package signing
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/stellar/go/keypair"
+	"github.com/stellar/go/network"
+	"github.com/stellar/go/txnbuild"
+	"github.com/stellar/wallet-backend/internal/db"
+	"github.com/stellar/wallet-backend/internal/signing/channelaccounts"
+)
+
+type channelAccountDBSignatureClient struct {
+	networkPassphrase    string
+	dbConnectionPool     db.ConnectionPool
+	encryptionPassphrase string
+	privateKeyEncrypter  channelaccounts.PrivateKeyEncrypter
+	channelAccountStore  channelaccounts.ChannelAccountStore
+}
+
+var _ SignatureClient = (*channelAccountDBSignatureClient)(nil)
+
+func NewChannelAccountDBSignatureClient(dbConnectionPool db.ConnectionPool, networkPassphrase string, privateKeyEncrypter channelaccounts.PrivateKeyEncrypter, encryptionPassphrase string) (*channelAccountDBSignatureClient, error) {
+	if networkPassphrase != network.TestNetworkPassphrase && networkPassphrase != network.PublicNetworkPassphrase {
+		return nil, fmt.Errorf("invalid network passphrase provided: %s", networkPassphrase)
+	}
+
+	return &channelAccountDBSignatureClient{
+		networkPassphrase:    networkPassphrase,
+		dbConnectionPool:     dbConnectionPool,
+		channelAccountStore:  channelaccounts.NewChannelAccountModel(dbConnectionPool),
+		privateKeyEncrypter:  privateKeyEncrypter,
+		encryptionPassphrase: encryptionPassphrase,
+	}, nil
+}
+
+func (sc *channelAccountDBSignatureClient) NetworkPassphrase() string {
+	return sc.networkPassphrase
+}
+
+func (sc *channelAccountDBSignatureClient) GetAccountPublicKey(ctx context.Context) (string, error) {
+	channelAccount, err := sc.channelAccountStore.GetIdleChannelAccount(ctx, time.Minute)
+	if err != nil {
+		return "", fmt.Errorf("getting idle channel account: %w", err)
+	}
+	return channelAccount.PublicKey, nil
+}
+
+func (sc *channelAccountDBSignatureClient) getKPsForPublicKeys(ctx context.Context, stellarAccounts ...string) ([]*keypair.Full, error) {
+	if len(stellarAccounts) == 0 {
+		return nil, fmt.Errorf("no accounts provided")
+	}
+
+	accountsAlreadyAccountedFor := map[string]struct{}{}
+	kps := []*keypair.Full{}
+	for i, account := range stellarAccounts {
+		if _, ok := accountsAlreadyAccountedFor[account]; ok {
+			continue
+		}
+		accountsAlreadyAccountedFor[account] = struct{}{}
+
+		if account == "" {
+			return nil, fmt.Errorf("account %d is empty", i)
+		}
+
+		chAcc, err := sc.channelAccountStore.Get(ctx, sc.dbConnectionPool, account)
```

Comment:
💡 if we want to make it more performant, perhaps we could get all accounts from the database at once, before the loop